### PR TITLE
New Module:  win_gpo_manage

### PR DESCRIPTION
**Pull request by sgruber94**
### SUMMARY
The wwin_gpo_manage module manage a GroupPolicy

### ISSUE TYPE
- New Module Pull Request

### COMPONENT NAME
win_gpo_manage

### ADDITIONAL INFORMATION
Includes Integration Tests - See documentation file for full details on how the module works.

### CODE REVIEW

File: plugins/modules/win_gpo_manage.ps1 [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+# Copyright: (c) 2021 Sebastian Gruber ,dacoso GmbH All Rights Reserved.
+# SPDX-License-Identifier: GPL-3.0-only
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+#AnsibleRequires -CSharpUtil Ansible.Basic
+
+$spec = @{
+    options  = @{
+        name     = @{ type = "list"; elements = "str" }
+        folder   = @{ type = "path"; default = "C:\GPO" }
+        mode     = @{ type = "str"; choices = "import", "query", "remove", "export";required=$true }
```
Comment (jborean93) [inline plugins/modules/win_gpo_manage.ps1:11]: Use `state` instead as it's a more common option in Ansible. It should have the following options

* `import` -> `present`
* `query` -> this shouldn't be here but in a separate `win_gpo_manage_info` module
* `remove` -> `absent`
* `export` -> `exported`

It should also have a default of `present` just like most of the other modules that have `state`.

File: plugins/modules/win_gpo_manage.ps1 [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+}
+$module = [Ansible.Basic.AnsibleModule]::Create($args, $spec)
+$check_mode = $module.CheckMode
+$gpomode = $module.Params.mode
+$removedgpo = $module.Params.name
+$folderpath = $module.Params.folder
+try {
+    Import-Module GroupPolicy
+} catch {
+    $module.FailJson("win_grouppolicy requires the GroupPolicy PS module to be installed")
```
Comment (jborean93) [inline plugins/modules/win_gpo_manage.ps1:29]:
```suggestion
    $module.FailJson("win_gpo_manage requires the GroupPolicy PS module to be installed", $_)
```

File: plugins/modules/win_gpo_manage.ps1 [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    }
+    $gpo = Get-GPO -All
+    foreach ($Entry in $gpo) {
+        $Path = $ExportFolder + "\" + $entry.Displayname
+        if((Test-Path -LiteralPath $Path) -and !($override)) {
+            $module.FailJson("Folder already exists, Override is false ")
+        } else {
+            if(!(Test-Path -LiteralPath $Path)) {
+                #Folder doesnt exist
+                New-Item -ItemType directory -Path $Path -WhatIf:$check_mode
```
Comment (jborean93) [inline plugins/modules/win_gpo_manage.ps1:79]: I'm pretty sure `New-Item` outputs an object, you need to pipe to Out-Null or assign it to a var so it doesn't impact our module output result.

File: plugins/modules/win_gpo_manage.ps1 [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        }
+    }
+}
+if ($diff_mode) {
+    $module.result.diff = @{}
+}
+if ($gpomode -eq "remove") {
+    #Removes GPOs
+    $gpo = (Get-GPO -All).Displayname
+    $module.diff.before = $gpo | Out-String
```
Comment (jborean93) [inline plugins/modules/win_gpo_manage.ps1:96]: No need for the `| Out-String`, the diff builder is able to compare a list in before and after and show what was added/removed.

File: plugins/modules/win_gpo_manage.ps1 [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            if((Test-Path -LiteralPath $Path) -and ($override)) {
+                Remove-Item -ItemType directory -Path $Path -WhatIf:$check_mode
+                New-Item -ItemType directory -Path $Path -WhatIf:$check_mode
+            }
+            Backup-GPO -Guid $Entry.id -Path $Path -WhatIf:$check_mode
+            $module.result.exported.Add($entry.Displayname)
+        }
+    }
+}
+if ($diff_mode) {
```
Comment (jborean93) [inline plugins/modules/win_gpo_manage.ps1:90]: You haven't defined `$diff_mode`, use `$module.DiffMode` to check if diff mode is enabled. Ultimately you don't even need to check. You can always set the diff settings and the controller is able to selectively display it if `--diff` was enabled. In fact I recommend you always set it so your code runs the diff stuff and won't randomly fail due to a diff specific setting when `--diff` is enabled.

File: plugins/modules/win_gpo_manage.ps1 [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    $module.result.diff = @{}
```
Comment (jborean93) [inline plugins/modules/win_gpo_manage.ps1:91]: You should use `$module.Diff` instead to set diff values. It's already pre-populated with an empty hashtable.

File: plugins/modules/win_gpo_manage.ps1 [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    if ($diff_mode -and $result.changed) {
+        if (!$check_mode) {
+            #only check
+            $module.result.diff.after = [Array]$removedgpo | Out-String
+        } else {
+            #remove gpos
+            $gpoafter = (Get-GPO -All).Displayname
+            $after = [System.Collections.Generic.List`1[String]]
+            $after.AddRange($gpoafter)
+            $module.result.diff.after = $after | Out-String
```
Comment (jborean93) [inline plugins/modules/win_gpo_manage.ps1:124]: `$after` is now a type and not the list you want it to be so running `$after.AddRange($gpoafter)` will not work. Ultimately you can simplify this to just

```suggestion
            $module.Diff.after = @((Get-Gpo -All).Displayname)`
```

File: plugins/modules/win_gpo_manage.ps1 [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        }
+    }
+}
+if ($gpomode -eq "query") {
```
Comment (jborean93) [inline plugins/modules/win_gpo_manage.ps1:128]: This should be in it's own `win_gpo_manage_info` module (really probably `win_gpo_info`.

File: plugins/modules/win_gpo_manage.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2021 Sebastian Gruber ,dacoso GmbH All Rights Reserved.
+# SPDX-License-Identifier: GPL-3.0-only
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+DOCUMENTATION = r'''
+---
+module: win_gpo_manage
```
Comment (jborean93) [inline plugins/modules/win_gpo_manage.py:9]: Needs `version_added: '1.4.0'`.

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: plugins/modules/win_gpo_manage.ps1 (hunk 8 above)
Comment (sgruber94) [reply plugins/modules/win_gpo_manage.ps1:125]: i will generate a new module for that

**Comment by briantist** [comment]
@sgruber94 it looks like you added a merge commit instead of rebasing your branch against main.. there's a ton of "changes" in here that shouldn't be here. I'm not entirely sure how to fix it, but it seems like you need a rebase, at a minimum.

**Comment by jborean93** [comment]
See https://github.com/ansible-collections/community.windows/pull/194#issuecomment-877442730

**Pull request closed without merging**